src/sink/dns.rs

```diff
@@ -0,0 +1,192 @@
+        RData::HINFO(v) => json_typed!(borrowed, { "HINFO": v.to_string() }).into(),
+        RData::HTTPS(v) => json_typed!(borrowed, { "HTTPS": v.to_string() }).into(),
+        RData::MX(v) => json_typed!(borrowed, { "MX": v.to_string() }).into(),
+        RData::NAPTR(v) => json_typed!(borrowed, { "NAPTR": v.to_string() }).into(),
+        RData::NULL(v) => json_typed!(borrowed, { "NULL": v.to_string() }).into(),
+        RData::NS(v) => json_typed!(borrowed, { "NS": v.to_string() }).into(),
+        RData::OPENPGPKEY(v) => json_typed!(borrowed, { "OPENPGPKEY": v.to_string() }).into(),
+        RData::SOA(v) => json_typed!(borrowed, { "SOA": v.to_string() }).into(),
+        RData::SRV(v) => json_typed!(borrowed, { "SRV": v.to_string() }).into(),
+        RData::SSHFP(v) => json_typed!(borrowed, { "SSHFP": v.to_string() }).into(),
+        RData::SVCB(v) => json_typed!(borrowed, { "SVCB": v.to_string() }).into(),
+        RData::TLSA(v) => json_typed!(borrowed, { "TLSA": v.to_string() }).into(),
+        RData::OPT(_) | RData::Unknown { .. } | RData::ZERO => return None,
+    })
+}
+fn lookup_to_value(l: &Lookup) -> Value<'static> {
+    l.iter().filter_map(rdata_to_value).collect()
+}
+
+#[async_trait::async_trait]
+impl Sink for Dns {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        _codec: &mut dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        event: Event,
+    ) -> ResultVec {
+        let resolver = self
+            .resolver
+            .as_ref()
+            .ok_or_else(|| Error::from("No resolver set"))?;
+        let mut res = Vec::with_capacity(event.len());
+        let correlation = event.correlation_meta();
+
+        for e in event.value_iter() {
+            if let Some(lookup) = e.get("lookup") {
+                if let Some(name) = lookup.as_str().or_else(|| lookup.get_str("name")) {
+                    let data = if let Some(record_type) =
+                        lookup.get_str("type").and_then(str_to_record_type)
+                    {
+                        // type lookup
+                        lookup_to_value(
+                            &resolver
+                                .lookup(name, record_type, DnsRequestOptions::default())
+                                .await?,
+                        )
+                    } else {
+                        // generic lookup
+                        lookup_to_value(resolver.lookup_ip(name).await?.as_lookup())
+                    };
+                    let e = Event {
+                        data: (data, correlation.clone().unwrap_or_default()).into(),
+                        origin_uri: Some(self.event_origin_uri.clone()),
+                        ..Event::default()
+                    };
+
+                    res.push(Reply::Response("out".into(), e))
+                } else {
+                    error!("Invalid DNS request: {}", e)
```

Comment:
I think we should add Error events in this case and in all other error cases too and send them out the `err` port.